src/App.php

```diff
@@ -661,42 +661,78 @@ public function url($page = [], bool $useRequestUrl = false, array $extraRequest
         $request = $this->getRequest();
 
         if ($useRequestUrl) {
-            $page = $request->getUri()->getPath();
-            $query = $request->getUri()->getQuery();
-            if ($query !== '') {
-                $page .= '?' . $query;
-            }
+            $page = $this->urlGetCurrentRequestPage();
         }
 
-        $pagePath = '';
         if (is_string($page)) {
-            $pageExploded = explode('?', $page, 2);
-            $pagePath = $pageExploded[0];
-            parse_str($pageExploded[1] ?? '', $page);
-        } else {
-            if (isset($page[0])) {
-                $pagePath = $page[0];
-            } else {
-                // use current page by default
-                $requestUrl = $request->getUri()->getPath();
-                if (substr($requestUrl, -1, 1) === '/') {
-                    $pagePath = $this->urlBuildingIndexPage;
-                } else {
-                    $pagePath = basename($requestUrl, $this->urlBuildingExt);
-                }
-            }
-            unset($page[0]);
-            if ($pagePath) {
-                $pagePath .= $this->urlBuildingExt;
-            }
+            $page = $this->urlSplitStringPageIntoArray($page);
         }
 
+        $pagePath = $this->urlConstructPagePath($page[0] ?? $request->getUri()->getPath());
+        unset($page[0]);
+
+        $args = $this->urlMergeArguments($page, $extraRequestUrlArgs);
+
+        $pageQuery = http_build_query($args, '', '&', \PHP_QUERY_RFC3986);
+
+        return $pagePath . ($pageQuery ? '?' . $pageQuery : '');
+    }
+
+    private function urlGetCurrentRequestPage(): array
+    {
+        $request = $this->getRequest();
+        $page = $request->getQueryParams();
+        $page[0] = $request->getUri()->getPath();
+
+        return $page;
+    }
+
+    private function urlSplitStringPageIntoArray(string $page): array
+    {
+        $pageExploded = explode('?', $page, 2);
+        $arrayPage = [];
+        parse_str($pageExploded[1] ?? '', $arrayPage);
+        $arrayPage[0] = $pageExploded[0];
+
+        return $arrayPage;
+    }
+
+    private function urlConstructPagePath(string $pagePath): string
+    {
+        $prefix = '';
+        if (substr($pagePath, 0, 2) === '..') {
+            $prefix = '..';
+        }
+
+        // Changed array string access to substr for PHP 7.4 compatibility
+        $lastChar = substr($pagePath, -1);
+        if ($lastChar === '/') {
+            return $prefix . $pagePath . $this->urlBuildingIndexPage . $this->urlBuildingExt;
+        }
+
+        if ($pagePath === '') {
+            return '/' . $pagePath . $this->urlBuildingIndexPage . $this->urlBuildingExt;
+        }
+
+        $pagePathPart = trim(dirname($pagePath), '.');
```

Comment:
this is super wrong, we cannot just trim unlimited `.`

also `dirname('/xxx')` on Windows results `/`

let me fix this

Reply:
@abbadon1334 please review the 4 last commits

The end changed LoC is minimal.

I have updated the PR title - is this what you wanted or something else/more?

Reply:
Switching title to "preserve" make sense even for tests without full path.
LGTM and thanks to go ahead on my work and do it better